aerogear/data-sync-server
# feat: initial subscriptions implementation

#### To do

* Tests
* Cleanup???

This is runnable right now though. You can try it out as follows:

* `npm run db:init:memeo`
* `npm run dev`
* Open `http://localhost:8000/graphiql` in two different tabs
* Leave one tab as it is with the sample queries
* In the other tab remove the sample queries and paste the following:

```
subscription memeAdded{
  memeAdded{
    id
    photoUrl
  }
}
```
* Run the subscription query and you should see the following message: `"Your subscription data will appear here after server publication!"`
* Run the createMeme query in the other window.

### How does the subscription piece work?

On the server side we define a new `memeAdded` subscription.

```
type Subscription {
    memeAdded(photoUrl: String):Meme!
}
```

Clients will be able to use this query to listen for updates. The photoUrl variable is optional (more on that later).

The next thing we need to be able to do is define a way to publish an event during some query or mutation. GraphQL uses a Pub/Sub model to do this. So when a query/mutation occurs, we can publish some events+data to a Pub/Sub mechanism and we can react to these events elsewhere. Below shows how a user could configure a `createMeme` resolver to also publish an event.

```js
{
    type: 'Mutation',
    field: 'createMeme',
    requestMapping: `{
      "operation": "insert",
      "doc": {
        "_type":"meme",
        "photoUrl": "{{context.arguments.photoUrl}}"
      }
    }`,
    responseMapping: '{{ toJSON (convertNeDBIds context.result) }}',
    publish: JSON.stringify({
      topic: 'memeCreated',
      payload: `{
        "memeAdded": {{ toJSON context.result }},
      }`
    }),
  }
```

The last piece of the puzzle is to be able to listen for those events and push them back down to a client. We can do this with the new `Subscription` config object. Take a look at the example one for the memeolist (included in this PR)

```js
{
    type: 'Subscription',
    field: 'memeAdded',
    topic: 'memeCreated',
}
```

This config tells our server to listen for messages published to the 'memeCreated' topic on the internal pubsub mechanism and to fire the `memeAdded` resolver as a result.

## Trying out Filtering

Filtering is an important aspect of Grapqhl subscriptions. In most real world apps, it doesn't make sense for a client to be listening for notifications about every single record that was added to a collection. It would be like getting a notification in slack every time anybody in the world sent a message.

Taking the Slack example further, on the client side we might subscribe to a messageAdded query as follows:

```
subscription messageAdded{
  messageAdded(channelId: '#random'){
    id
	channelId
    text
	userId
  }
}
```

The idea is we only want to listen for notifications in a particular slack channel - `#random`.

### How would this look on the server side?

How can we enable an end user of our service to define some kind of filtering that's relatively good?

I have come up with small DSL that allows the user to define some conditions in a **declarative** way using **JSON**. We add this filter to the **subscription** config object.

Here's when it might look like in the Slack example

```js
{
    type: 'Subscription',
    field: 'messageAdded',
    topic: 'messageCreated',
	filter: JSON.stringify({
      'eq': ['$payload.messageAdded.channelId', '$variables.channelId']
    })
  }
```

This filter would check that the channelId in the newly created message is equal to the channelId that the client passed as a variable when they subscribed. i.e. the user can say **give me real time updates about messages for this particular channel**.

The filter expression has access to a context object that looks like this: `{ payload: {...}, variables: {}}` where `payload` is the message received from the pubsub layer and variables contains whatever variables were passed by the client to initialize the subscription. Using the dollar syntax like `$payload.messageAdded.id` we can reference the context object.

I figured that 99% of the time, the basic use case will like the one above where we check that two things are equal. I made a more concise syntax for this particular case. The same example above can be expressed as follows:

```js
{
    type: 'Subscription',
    field: 'messageAdded',
    topic: 'messageCreated',
	filter: JSON.stringify({
      '$payload.messageAdded.channelId': '$variables.channelId'
    })
  }
```

This is the most basic usage of the filter language. You can do many more things. Try some of these out with the memeolist schema:

(you can modify the subscription object in the file located under `seeders`, then run the **db init and start commands again**)


```js
// subscribe only to new memes where payload.photoURL === google.com
{
    type: 'Subscription',
    field: 'memeAdded',
    GraphQLSchemaId: 2,
    topic: 'memeCreated',
    createdAt: time,
    updatedAt: time,
	filter: JSON.stringify({
      'eq': ['$payload.memeAdded.photoUrl', 'https://google.com']
    })
  }
```

```js
// subscribe only to new memes where payload.photoURL !== google.com
{
    type: 'Subscription',
    field: 'memeAdded',
    GraphQLSchemaId: 2,
    topic: 'memeCreated',
    createdAt: time,
    updatedAt: time,
	filter: JSON.stringify({
      '!eq': ['$payload.memeAdded.photoUrl', 'https://google.com']
    })
  }
```

```js
// subscribe only to new memes where payload.photoURL matches a regex
{
    type: 'Subscription',
    field: 'memeAdded',
    GraphQLSchemaId: 2,
    topic: 'memeCreated',
    createdAt: time,
    updatedAt: time,
	filter: JSON.stringify({
      'match': ['$payload.memeAdded.photoUrl', 'https:\/\/.*']
    })
  }
```

```js
// subscribe only to new memes where payload.photoUrl === https://google.com || payload.photoUrl == $variables.photoUrl
{
    type: 'Subscription',
    field: 'memeAdded',
    GraphQLSchemaId: 2,
    topic: 'memeCreated',
    createdAt: time,
    updatedAt: time,
	filter: JSON.stringify({
      'or': [
        { 'eq': ['$payload.memeAdded.photoUrl', 'https://google.com'] },
        { 'eq': ['$payload.memeAdded.photoUrl', '$variables.photoUrl']}
       ]
    })
  }
```

```js
// subscribe to new memes where payload.photoUrl === variables.photoUrl || (payload.photoUrl === https://facebook.com && payload.photoUrl === Dara)
{
    type: 'Subscription',
    field: 'memeAdded',
    GraphQLSchemaId: 2,
    topic: 'memeCreated',
    createdAt: time,
    updatedAt: time,
	filter: JSON.stringify({
      'or': [
        { 'eq': ['$payload.memeAdded.photoUrl', '$variables.photoUrl']},
        { 'and': [
                  { 'eq': ['$payload.memeAdded.photoUrl', 'https://facebook.com']},
                  { 'eq': ['$payload.memeAdded.owner', 'Dara'] }
                ]
       ]
    })
  }
```

### Why Did I do it this way?

* JSON format is **super portable**, for example - instead of defining the filter expression in the server config, a client app could define whatever filter they want and use it to listen for updates, **and it would be completely safe** because there's no way for them to inject code. This has the potential for some very useful applications IMO.
* I considered using handlebars to do something similar but it's very hacky because it doesn't actually evaluate things to `true` or `false`. It would also mean users would have to write a horrible syntax. In my opinion, this syntax is pretty intuitive.
* Inspired by MongoDB query language which I thought was very intuitive
* Flexibility, the code that evaluates these expressions could be extended in the future very easily.
